src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -201,37 +193,22 @@ internal override CommonAnonymousTypeManager CommonAnonymousTypeManager
         /// </summary>
         internal bool IsPeVerifyCompatEnabled => LanguageVersion < LanguageVersion.CSharp7_2 || Feature("peverify-compat") != null;
 
-        /// <summary>
-        /// Returns true if nullable analysis is enabled in this compilation.
-        /// </summary>
-        /// <remarks>
-        /// Returns false if analysis is explicitly disabled for all methods; otherwise
-        /// returns true if the nullable context is enabled in the compilation options
-        /// or if there are any nullable-enabled contexts in the compilation.
-        /// </remarks>
-        internal bool IsNullableAnalysisEnabled
+        internal bool IsNullableAnalysisEnabledIn(SyntaxNode syntax)
         {
-            get
-            {
-                if (!_lazyShouldRunNullableAnalysis.HasValue())
-                {
-                    _lazyShouldRunNullableAnalysis = isEnabled().ToThreeState();
-                }
-                return _lazyShouldRunNullableAnalysis.Value();
+            return IsNullableAnalysisEnabledIn((CSharpSyntaxTree)syntax.SyntaxTree, syntax.Span);
+        }
 
-                bool isEnabled()
-                {
-                    if (GetNullableAnalysisValue() == false)
-                    {
-                        return false;
-                    }
-                    if (Options.NullableContextOptions != NullableContextOptions.Disable)
-                    {
-                        return true;
-                    }
-                    return SyntaxTrees.Any(tree => ((CSharpSyntaxTree)tree).HasNullableEnables());
-                }
-            }
+        internal bool IsNullableAnalysisEnabledIn(CSharpSyntaxTree tree, TextSpan span)
+        {
+            return GetNullableAnalysisValue() ??
+                tree.IsNullableAnalysisEnabled(span) ??
+                (Options.NullableContextOptions & NullableContextOptions.Warnings) != 0;
+        }
+
+        internal bool IsNullableAnalysisEnabledIn(MethodSymbol method)
```

Comment:
Is the expectation that for any `SyntaxNode` in a given `MethodSymbol` that I should get equivalent results for the two overloads of `IsNullableAnalysisEnabledIn`? Basically is there a reason we should prefer say the symbol version over the syntax one in our code base or are they interchangable.

Reply:
The overload that takes a `SyntaxNode` is used to calculate the actual value for a specific span, typically during binding, including cases where the span is a method declaration (see `SourceMemberContainerSymbol.AddNonTypeMembers()`). The overload that takes a `MethodSymbol` is for callers that are typically deal with symbols rather than syntax and that overload delegates to the method that cached the value (possibly aggregated from several spans) calculated from the first overload.

---
In reply to: [551608662](https://github.com/dotnet/roslyn/pull/49953#discussion_r551608662) [](ancestors = 551608662)

Reply:
Consider the following method:

```c#
#nullable enable 
void M() {
  int local1 = 42;
#nullable disable
  int local2 = 13;
}
```

Is the expectation then that the `SyntaxNode` for `local1` will return `true` here but the one for `local2` will return `false`?